src/compiler/sys/in-memory-fs.ts

```diff
@@ -2,76 +2,225 @@ import type * as d from '../../declarations';
 import { basename, dirname, relative } from 'path';
 import { isIterable, normalizePath, isString } from '@utils';
 
+/**
+ * An in-memory FS which proxies the underlying OS filesystem using a simple
+ * in-memory cache. FS writes can accumulate on the in-memory system, using an
+ * API similar to Node.js' `"fs"` module, and then be committed to disk as a
+ * unit.
+ *
+ * Files written to the in-memory system can be edited, deleted, and so on.
+ * This allows the compiler to proceed freely as if it is modifying the
+ * filesystem, modifying the world in whatever way suits it, while deferring
+ * actual FS writes until the end of the compilation process, making actual
+ * changes to the filesystem on disk contingent on an error-free build or any
+ * other condition.
+ *
+ * Usage example:
+ *
+ * ```ts
+ * // create an in-memory FS
+ * const sys = createSystem();
+ * const inMemoryFs = createInMemoryFs(sys);
+ *
+ * // do a few fs operations
+ * await inMemoryFs.writeFile("path/to/file.js", 'console.log("hey!");')
+ * await inMemoryFs.remove("path/to/another_file.ts");
+ *
+ * // commit the results to disk
+ * const commitStats = await inMemoryFs.commit();
+ * ```
+ *
+ * In the above example the write operation and the delete operation (w/
+ * `.remove`) are both queue in the in-memory proxy but not committed to
+ * disk until the `.commit` method is called.
+ */
+export type InMemoryFileSystem = ReturnType<typeof createInMemoryFs>;
+
+/**
+ * An node in the in-memory file system. This may represent a file or
+ * a directory, and pending copy, write, and delete operations may be stored
+ * on it.
+ */
+export interface FsItem {
```

Comment:
some of these other interfaces are used outside of this file (although generally not too often). I considered moving them all to the bottom of the file below `createInMemoryFs` but didn't so yet. If that's a cleaner look I'm open to it!